DOC Improve User Guide for metadata routing (#27282)
scikit-learn/scikit-learn: doc/metadata_routing.rst

  enable it by setting the ``enable_metadata_routing`` flag to
  ``True``::

    >>> import sklearn
    >>> sklearn.set_config(enable_metadata_routing=True)

This guide demonstrates how :term:`metadata` can be routed and passed between objects in
scikit-learn.

Metadata is data that an estimator, scorer, or CV splitter takes into account if the
user explicitly passes it as a parameter. For instance, some methods in objects have
been implemented to use `sample_weight`, `classes`  or `groups` for computation if the
user deviates from the default `None`. If these objects are used in conjunction with
other objects, in the past there was no single API for passing the metadata around.

With the Metadata Routing API, we can transfer metadata to estimators, scorers, and CV
splitters using :term:`meta-estimators` (such as :class:`~pipeline.Pipeline` or
:class:`~model_selection.GridSearchCV`) or routing functions such as
:func:`~model_selection.cross_validate`. In order to pass metadata to a method such as
``fit`` or ``score``, the object consuming the metadata, must *request* it. This is done
via `set_{method}_request()` methods, where `{method}` is substituted by the name of the

> method that request the metadata. For instance, estimators and splitters, that use the
> metadata in their `fit()` method would use `set_fit_request()`, and scorers would use
> the `set_score_request()`.

Reviewer: for splitters it would have been `set_split_request`, but we have it the default to be requested for `groups` in our splitters, so rather not mention it here to avoid more confusion.

```suggestion
method that request the metadata. For instance, estimators that use the metadata in
their `fit()` method would provide a `set_fit_request()`, and scorers would use the
`set_score_request()`.
```

  enable this feature  by setting the ``enable_metadata_routing`` flag to
  enable it by setting the ``enable_metadata_routing`` flag to
  ``True``::

    >>> import sklearn
    >>> sklearn.set_config(enable_metadata_routing=True)

This guide demonstrates how metadata such as ``sample_weight`` can be routed
and passed along to estimators, scorers, and CV splitters through
meta-estimators such as :class:`~pipeline.Pipeline` and
:class:`~model_selection.GridSearchCV`. In order to pass metadata to a method
such as ``fit`` or ``score``, the object consuming the metadata, must *request*
it. For estimators and splitters, this is done via ``set_*_request`` methods,
e.g. ``set_fit_request(...)``, and for scorers this is done via the
``set_score_request`` method. For grouped splitters such as
:class:`~model_selection.GroupKFold`, a ``groups`` parameter is requested by
default. This is best demonstrated by the following examples.

If you are developing a scikit-learn compatible estimator or meta-estimator,
you can check our related developer guide:
:ref:`sphx_glr_auto_examples_miscellaneous_plot_metadata_routing.py`.

> .. note::

Reviewer: why remove the `.. note::` directive?

Author: Because I merged that note with the other one on top. It's cleaner to read like that.

Reviewer: Then you need to remove the indentation here.

Author: Hm, I cannot see why you suggest that.
I made those two notes into a larger one in the beginning of the document. All of that is supposed to be one note:

```
.. note::
  The Metadata Routing API is experimental, and is not implemented yet for some
  estimators. Please refer to the :ref:`list of supported and unsupported
  models <metadata_routing_models>` for more information. It may change without
  the usual deprecation cycle. By default this feature is not enabled. You can
  enable it by setting the ``enable_metadata_routing`` flag to
  ``True``::

    >>> import sklearn
    >>> sklearn.set_config(enable_metadata_routing=True)

  Note that the methods and requirements introduced in this document are only
  relevant if you want to pass :term:`metadata` (e.g. ``sample_weight``) to a method.
  If you're only passing ``X`` and ``y`` and no other parameter / metadata to
  methods such as :term:`fit`, :term:`transform`, etc., then you don't need to set
  anything.
```

If I remove the indentation from the second paragraph, then it's not included in the note anymore.

With the Metadata Routing API, we can transfer metadata to estimators, scorers, and CV
splitters using :term:`meta-estimators` (such as :class:`~pipeline.Pipeline` or
:class:`~model_selection.GridSearchCV`) or routing functions such as
:func:`~model_selection.cross_validate`. In order to pass metadata to a method like
``fit`` or ``score``, the object consuming the metadata, must *request* it. This is done
via `set_{method}_request()` methods, where `{method}` is substituted by the name of the

> method that request the metadata. For instance, estimators that use the metadata in

Reviewer 2: ```suggestion
method that requests the metadata. For instance, estimators that use the metadata in
```

> method that request the metadata. For instance, estimators and splitters, that use the

Reviewer: splitters don't have `fit`, they use `split`

Author: Alright, changed. Thanks.

> their `fit()` method would use `set_fit_request()`, and scorers would use the

Reviewer 2: ```suggestion
their `fit()` method would use `set_fit_request()`, and scorers would use
```

> the `set_score_request()`. These methods allow us to specify which metadata to request,

Reviewer: ```suggestion
the `set_score_request()`. These methods allow you to specify which metadata to request,
```

Author: If we use "we" then we also need to use "us".

instance `set_fit_request(sample_weight=True)`.

For grouped splitters such as :class:`~model_selection.GroupKFold`, a
``groups`` parameter is requested by default. This is best demonstrated by the
following examples.

Usage Examples
**************

> Here we present a few examples to show different common use-cases. Our goal is to pass

Reviewer 2: ```suggestion
Here we present a few examples to show some common use-cases. Our goal is to pass
```

`sample_weight` and `groups` through :func:`~model_selection.cross_validate`, which
routes the metadata to :class:`~linear_model.LogisticRegressionCV` and to a custom scorer

> made with :func:`~metrics.make_scorer` that both *can* use the metadata in their

Reviewer 2: ```suggestion
made with :func:`~metrics.make_scorer`, both of which *can* use the metadata in their
```

    >>> import sklearn
    >>> sklearn.set_config(enable_metadata_routing=True)

This guide demonstrates how :term:`metadata`, such as ``sample_weight`` can be routed
and passed between objects in scikit-learn. Metadata can be transfered to estimators,
scorers, and CV splitters using :term:`meta-estimators` such as
:class:`~pipeline.Pipeline` or :class:`~model_selection.GridSearchCV`.  It can also be

> routed between these objects without the presence of any meta-estimator. In order to

Reviewer: what do you mean by 

> It can also be
routed between these objects without the presence of any meta-estimator.

Author: I added an explicit mention of functions that are able to route.

pass metadata to a method such as ``fit`` or ``score``, the object consuming the
metadata, must *request* it. This is done via `set_{method}_request()` methods. For
instance, estimators and splitters would use `set_fit_request()`, and scorers would use

> the `set_score_request()` method. The following methods are supported:

Reviewer: The following methods are not _really_ supported. Most meta-estimators don't support passing alone metadata other than for `fit` and `score`, and some for `split`. Everything else will be a long road ahead.

`fit`,
`fit_predict`,
`fit_transform`,
`partial_fit`,
`transform`,
`inverse_transform`,
`predict`,
`predict_log_proba`,
`predict_proba`,
`decision_function`,
`score`,

> and `split`.

Reviewer: you can put these in a bullet list

Author: I've deleted this list, because it's only a list of methods some objects in scikit-learn have right now. It's better to mention most common methods instead and hint that there are just examples.

Weighted scoring and fitting
----------------------------

The splitter used internally in :class:`~linear_model.LogisticRegressionCV`,
:class:`~model_selection.GroupKFold`, requests ``groups`` by default. However, we need
to explicitly request `sample_weight` for it and for our custom scorer by specifying
`sample_weight=True` in :class:`~linear_model.LogisticRegressionCV`s `set_fit_request()`
method and in :func:`~metrics.make_scorer`s `set_score_request()` method. Both
:term:`consumers <consumer>` know how to use ``sample_weight`` in their `fit()` or
`score()` methods. We can then pass the metadata in

> :func:`~model_selection.cross_validate` that will route it to any active consumers.::

Reviewer: ```suggestion
:func:`~model_selection.cross_validate` that will route it to any active consumers::
```

>>> weighted_acc = make_scorer(accuracy_score).set_score_request(sample_weight=True)
>>> lr = LogisticRegressionCV(
...     cv=GroupKFold(),
...     scoring=weighted_acc
... ).set_fit_request(sample_weight=True)
>>> cv_results = cross_validate(
...     lr,
...     X,
...     y,
...     params={"sample_weight": my_weights, "groups": my_groups},
...     cv=GroupKFold(),
...     scoring=weighted_acc,
... )

Note that in this example, :func:`~model_selection.cross_validate` routes ``my_weights``

> to both the scorer and :class:`~linear_model.LogisticRegressionCV`.

Author: Let's explicitly name the `UnsetMetadataPassedError` then. The next sentence about error handling was already referring to it.

It's not that trivial though. If you only forget to `set_score_request(sample_weight=True)` on the scorer, you won't get an error message (since your `sample_weight` has been used somewhere). There is no way to know where a user intended to use the routed metadata, except for when they forget to use the `set_{method}_requests` entirely. Not routing it to the scorer could be done intentionally.

I tried to put this in words, so that it's simple enough for that introductory section and still conveys how that error works.

Weighted scoring and fitting
----------------------------

:class:`~model_selection.GroupKFold` requests ``groups`` by default. However, we need to
explicitly request weights for our scorer and the internal cross validation of

> :class:`~linear_model.LogisticRegressionCV` by specifying `sample_weight=True`. Both of
> these :term:`consumers <consumer>` know how to use metadata called ``sample_weight``::

Reviewer: ```suggestion
:class:`~linear_model.LogisticRegressionCV` by specifying `sample_weight=True`. Both
:term:`consumers <consumer>` know how to use metadata called ``sample_weight``::
```

  >>> weighted_acc = make_scorer(accuracy_score).set_score_request(sample_weight=True)
  >>> lr = LogisticRegressionCV(cv=GroupKFold(), scoring=weighted_acc,)

>   >>> lr.set_fit_request(sample_weight=True)

Reviewer: why the format change? it's fine as is, isn't it?

Author: I had hoped for the longest possible lines, but now I got aware that black is not formatting rst files.

Though, I still think the old formatting it's complicated to read. Let me try some compromise and please let me know what you think.

  >>> cv_results = cross_validate(
  ...     lr,
  ...     X,
  ...     y,
  ...     params={"sample_weight": my_weights, "groups": my_groups},
  ...     cv=GroupKFold(),
  ...     scoring=weighted_acc,
  ... )

Note that in this example, ``my_weights`` is passed to both the scorer and
:class:`~linear_model.LogisticRegressionCV`.

Error handling: if ``params={"sample_weigh": my_weights, ...}`` were passed
(note the typo), :func:`~model_selection.cross_validate` would raise an error,
since ``sample_weigh`` was not requested by any of its underlying objects.

Weighted scoring and unweighted fitting
---------------------------------------

When passing metadata such as ``sample_weight`` around, all ``sample_weight``
:term:`consumers <consumer>` require weights to be either explicitly requested or not

> requested (i.e. ``True`` or ``False``). Thus, to perform an unweighted fit, we need to

Reviewer: the "when used in another
:term:`router` such as a :class:`~pipeline.Pipeline` or a ``*GridSearchCV``" is rather important, why remove it? Consumers don't need to request anything if they're not used in another router.

Reviewer: the significance of that part here is that the user doesn't need to set any requests if they are not going to put those estimators and scorers in a router. Users can always do `LogisticRegression().fit(X, y, sample_weight)` w/o setting any requests. How do you want to convey that message here?

Author: I was thinking of the "passing metadata round" phrase as passing it into a router.

What about:
```
When passing metadata such as ``sample_weight`` into a :term:`router`
(:term:`meta-estimators` or routing function), all ``sample_weight`` :term:`consumers
<consumer>` require weights to be either explicitly requested or explicitly not
requested (i.e. ``True`` or ``False``).
```

configure :class:`~linear_model.LogisticRegressionCV` to not request sample weights, so
that :func:`~model_selection.cross_validate` does not pass the weights along::

  >>> weighted_acc = make_scorer(accuracy_score).set_score_request(sample_weight=True)
  >>> lr = LogisticRegressionCV(cv=GroupKFold(), scoring=weighted_acc,)

>   >>> lr.set_fit_request(sample_weight=False)

Reviewer: same here

  >>> cv_results = cross_validate(
  ...     lr,
  ...     X,
  ...     y,
  ...     cv=GroupKFold(),
  ...     params={"sample_weight": my_weights, "groups": my_groups},
  ...     scoring=weighted_acc,
  ... )

If :meth:`linear_model.LogisticRegressionCV.set_fit_request` had not been called,
:func:`~model_selection.cross_validate` would raise an error because ``sample_weight``
is passed but :class:`~linear_model.LogisticRegressionCV` would not be explicitly
configured to recognize the weights.

Unweighted feature selection
----------------------------

Setting request values for metadata is only required if the object, e.g. estimator,

> scorer, etc., is a potential :term:`consumers <consumer>` of that metadata. Unlike

Reviewer: not just "potential", they need to set it if they are its consumer and the metadata is passed. Otherwise an error will be raised.

  ... )

Note that in this example, ``my_weights`` is passed to both the scorer and
:class:`~linear_model.LogisticRegressionCV` via the routing function

> :func:`~model_selection.cross_validate`.

Reviewer: Maybe write this as:

```
Note that in this example, :func:`~model_selection.cross_validate` routes ``my_weights``
to both the scorer and :class:`~linear_model.LogisticRegressionCV`.
```

  ...     scoring=weighted_acc,
  ... )

If :meth:`linear_model.LogisticRegressionCV.set_fit_request` had not been called,
:func:`~model_selection.cross_validate` would raise an error because ``sample_weight``
is passed but :class:`~linear_model.LogisticRegressionCV` would not be explicitly
configured to recognize the weights.

Unweighted feature selection
----------------------------

Routing metadata is only possible if the object's method knows how to use the metadata,
meaning they have to have it as a parameter. Only then we can set request values for
metadata using `set_fit_request(sample_weight=True)`, for instance. This makes the

> object a :term:`consumer <consumer>`.

Reviewer: strictly speaking this is not true. Estimators can configure having a parameter as a consumer w/o having it as an explicit arg in the signature of the method. But I don't mind having this as you changed if it's more clear.

Reviewer: Yes, they could be accepting `**kwarg` and explicitly setting request values for some of those args for the machinery to work. They don't have to be explicit parameters.

  ...     scoring=weighted_acc,
  ... )

If :meth:`linear_model.LogisticRegressionCV.set_fit_request` had not been called,
:func:`~model_selection.cross_validate` would raise an error because ``sample_weight``
is passed but :class:`~linear_model.LogisticRegressionCV` would not be explicitly

> configured to recognize the weights.

Reviewer: That's an interesting suggestion, feel free to open a separate issue for it.

> configured to recognize the weights.

Author: Well, it seems that the error message is already stating this information, except for `cross_validate` for specific reasons of it's implementation. Here, it gets substituted by a changed version:

sklearn/model_selection/_validation.py
```python
        except UnsetMetadataPassedError as e:
            # The default exception would mention `fit` since in the above
            # `process_routing` code, we pass `fit` as the caller. However,
            # the user is not calling `fit` directly, so we change the message
            # to make it more suitable for this case.
            raise UnsetMetadataPassedError(
                message=(
                    f"{sorted(e.unrequested_params.keys())} are passed to cross"
                    " validation but are not explicitly requested or unrequested. See"
                    " the Metadata Routing User guide"
                    " <https://scikit-learn.org/stable/metadata_routing.html> for more"
                    " information."
                ),
                unrequested_params=e.unrequested_params,
                routed_params=e.routed_params,
            )
```
I don't think this is so easy to change.

Maybe, I could modify the error message that is checked in `sklearn/tests/test_metaestimators_metadata_routing.py,` though, because it's only considering the usual case...

> configured to recognize the weights.

Reviewer: That test file only checks for meta-estimators, and `cross_validate` is not a meta estimator. But I'm happy to make sure `cross_validate` also raises a nice error.

  ...     scoring=weighted_acc,
  ... )

If :meth:`linear_model.LogisticRegressionCV.set_fit_request` had not been called,
:func:`~model_selection.cross_validate` would raise an error because ``sample_weight``
is passed but :class:`~linear_model.LogisticRegressionCV` would not be explicitly

> configured to recognize the weights.

Author: The user is supposed to know what metadata their scorer (and their other objects used via `cross_validate`) is able to consume and what they want to route. This scorer cannot consume `groups` nor can `LogisticRegressionCV`.

And only metatdata that their objects can consume need to be set as requested or not requested. If they forget to set a metadata as requested or unrequested, they get an error message that explicitly states, what and where.

I agree that there is quite a threshold to enable people to use metadata routing.

I wonder if adding 2-3 sentences right in the beginning of the usage example section stating that users need to be clear which metadata can be consumed by their objects and have a little "plan" helps with this, or if it only adds clutter. After all, most users would not have the goal "user metadata routing" in mind, but rather "how can i make sure the same sample_weight is used everywhere" kind of question, and then find metadata routing as a solution.

  ...     scoring=weighted_acc,
  ... )

If :meth:`linear_model.LogisticRegressionCV.set_fit_request` had not been called,
:func:`~model_selection.cross_validate` would raise an error because ``sample_weight``
is passed but :class:`~linear_model.LogisticRegressionCV` would not be explicitly
configured to recognize the weights.

Unweighted feature selection
----------------------------

> Routing metadata is only possible if the object's method knows how to use the metadata,
> meaning they have to have it as a parameter. Only then we can set request values for

Reviewer: ```suggestion
Routing metadata is only possible if the object's method knows how to use the metadata,
which in most cases means they have it as an explicit parameter. Only then we can set request values for
```

metadata using `set_fit_request(sample_weight=True)`, for instance. This makes the
object a :term:`consumer <consumer>`.

Unlike :class:`~linear_model.LogisticRegressionCV`,
:class:`~feature_selection.SelectKBest` can't consume weights and therefore no request
value for ``sample_weight`` on its instance is set and ``sample_weight`` is not routed
to it::

  >>> weighted_acc = make_scorer(accuracy_score).set_score_request(sample_weight=True)
  >>> lr = LogisticRegressionCV(

>   ...     cv=GroupKFold(), scoring=weighted_acc

Reviewer: that comma at the end can stay there, there's no issue with that.

``sample_weight`` in ``fit`` and ``score``, it exposes
``estimator.set_fit_request(sample_weight=value)`` and
``estimator.set_score_request(sample_weight=value)``. Here ``value`` can be:

- ``True``: method requests a ``sample_weight``. This means if the metadata is provided,
  it will be used, otherwise no error is raised.
- ``False``: method does not request a ``sample_weight``.
- ``None``: router will raise an error if ``sample_weight`` is passed. This is in almost
  all cases the default value when an object is instantiated and ensures the user sets
  the metadata requests explicitly when a metadata is passed. The only exception are
  ``Group*Fold`` splitters.
- ``"param_name"``: alias for ``sample_weight`` if we want to pass different weights to
  different consumers. If aliasing is used the meta-estimator should not forward
  ``"param_name"`` to the consumer, but ``sample_weight`` instead, because the consumer
  will expect a param called ``sample_weight``. This means the mapping between the
  metadata required by the object, e.g. ``sample_weight`` and the variable name provided
  by the user, e.g. ``my_weights`` is done at the router level, and not by the consuming

>   object itself.

Author: And I would suggest to remove the part (starting from "_If aliasing ...."_) from here and add it to the example (examples/miscellaneous/plot_metadata_routing.py). The reason for this being is that people who only want to use metadata routing don't need to be burdened with this information.

I'd do that in PR #27357. Is this okay?

Reviewer: No I rather keep it here for the same of completeness of the API section.

``estimator.set_score_request(sample_weight=value)``. Here ``value`` can be:

- ``True``: method requests a ``sample_weight``. This means if the metadata is provided,
  it will be used, otherwise no error is raised.
- ``False``: method does not request a ``sample_weight``.
- ``None``: router will raise an error if ``sample_weight`` is passed. This is in almost
  all cases the default value when an object is instantiated and ensures the user sets
  the metadata requests explicitly when a metadata is passed. The only exception are
  ``Group*Fold`` splitters.
- ``"param_name"``: alias for ``sample_weight`` if we want to pass different weights to
  different consumers. If aliasing is used the meta-estimator should not forward
  ``"param_name"`` to the consumer, but ``sample_weight`` instead, because the consumer
  will expect a param called ``sample_weight``. This means the mapping between the
  metadata required by the object, e.g. ``sample_weight`` and the variable name provided
  by the user, e.g. ``my_weights`` is done at the router level, and not by the consuming
  object itself.

Metadata are requested in the same way for scorers using ``set_score_request``.

> If a metadata, e.g. ``sample_weight``, is passed by the user, the metadata request for

Reviewer 3: I've been wondering what word to use to refer to one of the entries in the dictionary that is passed as `params=`.

I think metadata is like data in that there is no plural. "The data show ...", "The author, title and page number are contained in the metadata". This means "a metadata" reads weird. How about "If an item of metadata, e.g. `sample weight`, is present in the metadata, the requests for all objects which can consume `sample_weight` should be configured by the user" (I also made some other changes to the sentence, just suggestions).

"an item of metadata" is the best I could come up with, I like it because in Python one entry in a dictionary (key and value) is already known as an item. This means the whole dictionary `{"sample_weight: ..., "groups": ..}` would be "the metadata" and one item of it (e.g. the weights) would be "an item of metadata".

You could argue that the weights by themselves are metadata, but then we need a term for the weights and groups together. Maybe "the set of metadata", but I like that less than my proposal above.

WDYT?

Author: What about "some metadata" (and then use plural)?

"set of" or "item of" both make me wonder what exactly is meant. And maybe we don't need to distinguish different quantities here.

  in almost all cases the default value when an object is instantiated and
  ensures the user sets the metadata requests explicitly when a metadata is
  passed. The only exception are ``Group*Fold`` splitters.
- ``"param_name"``: alias for ``sample_weight`` if we want to pass different weights to
  different consumers via `fit` and `score`. If aliasing is used in a meta-estimator,

>   the meta-estimator should forward ``"param_name"`` as ``sample_weight`` to the

Reviewer: aliasing is always mapped and handled in meta-estimator levels, when is it not like that?

  ``my_weights`` is done at the router level, and not by the object, e.g.
  estimator, itself.

> Metadata are requested in the same way for scorers using ``set_score_request``.

Reviewer: This is important and not covered in the above list, why remove it?

Author: Yes, but you already wrote this in the beginning of the list:

```
For instance, if an estimator supports ``sample_weight`` in ``fit`` and ``score``, it exposes 
``estimator.set_fit_request(sample_weight=value)`` and ``estimator.set_score_request(sample_weight=value)``.
```

And the list is general enough to apply to both methods.

Reviewer: scorers don't have a score method, they're a callable, that's why they're different.

Author: Okay, I didn't know that. But it seems to me that doesn't make a difference for users.

Could you please have a look at lines 208-211:
```
 For instance, if an estimator supports
``sample_weight`` in ``fit`` and ``score``, it exposes
``estimator.set_fit_request(sample_weight=value)`` and
``estimator.set_score_request(sample_weight=value)``. 
```
I think `set_score_request` is already covered there. So if we keep the line:
`Metadata are requested in the same way for scorers using ``set_score_request``.`
then should we exclude the scoring from above?

Reviewer: >  Metadata are requested in the same way for scorers using ``set_score_request``.

Yes, that statement covers it, but you have removed it here.

Author: Sorry, @adrinjalali, I meant to include it back yesterday, but I've just seen that I forgot to push. I didn't mean to make it difficult.
